File: Solutions-Python/two_sum.py
# time: O(n^2)
# space: O(1)
def two_sum_brute(nums, target):
    for i in range(len(nums) - 1):
        for j in range(i+1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]


# time: O(n)
# space: O(n)
def two_sum_hash(nums, target):
    nums_dict = {}
    for i in range(len(nums)):
        nums_dict[nums[i]] = i
    for i in range(len(nums)):
        complement = target - nums[i]
        if complement in nums_dict and nums_dict[complement] != i:
            return [i, nums_dict[complement]]


def my_print(nums, target, result):
    print("nums:", nums)
    print("target:", target)
    print("result:", result, "\n")


if __name__ == '__main__':
    print("Brute force approach...")
    nums1 = [2, 7, 11, 15]
    target1 = 9
    result = two_sum_brute(nums1, target1)
    my_print(nums1, target1, result)

    nums2 = [3, 2, 4]
    target2 = 6
    result = two_sum_brute(nums2, target2)
    my_print(nums2, target2, result)

    nums3 = [3, 3]
    target3 = 6
    result = two_sum_brute(nums3, target3)
    my_print(nums3, target3, result)

    print("Hash map approach...")
    result = two_sum_hash(nums1, target1)
    my_print(nums1, target1, result)

    result = two_sum_hash(nums2, target2)
    my_print(nums2, target2, result)

    result = two_sum_hash(nums3, target3)
    my_print(nums3, target3, result)

File: Solutions-Java/src/main/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class TwoSum {

    public static void main(String[] args) {
        System.out.println("Hello! Java!");

        // Setting up input
        int[] nums1 = {2, 7, 11, 15};
        int target1 = 9;

        int[] nums2 = {3, 2, 4};
        int target2 = 6;

        int[] nums3 = {3, 3};
        int target3 = 6;

        System.out.println("Brute force approach...");
        int[] result;
        result = twoSumBrute(nums1, target1);
        myPrint(nums1, target1, result);

        result = twoSumBrute(nums2, target2);
        myPrint(nums2, target2, result);

        result = twoSumBrute(nums3, target3);
        myPrint(nums3, target3, result);

        System.out.println("Hash map approach...");
        result = twoSumHash(nums1, target1);
        myPrint(nums1, target1, result);

        result = twoSumHash(nums2, target2);
        myPrint(nums2, target2, result);

        result = twoSumHash(nums3, target3);
        myPrint(nums3, target3, result);
    }

    // Time: O(n^2)
    // Space: O(1)
    public static int[] twoSumBrute(int[] nums, int target) {
        for (int i = 0; i < nums.length - 1; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    return new int[] {i, j};
                }
            }
        }

        return new int[] {-1};
    }

    // Time: O(n)
    // Space: O(n)
    public static int[] twoSumHash(int[] nums, int target) {
        Map<Integer, Integer> numsMap = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            numsMap.put(nums[i], i);
        }

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (numsMap.containsKey(complement) && numsMap.get(complement) != i) {
                return new int[] {i, numsMap.get(complement)};
            }
        }

        return new int[] {-1};
    }

    public static void myPrint(int[] nums, int target, int[] result) {
        System.out.println("nums = " + Arrays.toString(nums));
        System.out.println("target = " + target);
        System.out.println("result = " + Arrays.toString(result) + "\n");
    }
}
